src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -127,6 +162,73 @@ def getTaskData(environment_name, module_name, task_name)
 
   private
 
+  # @return Puppet::Node::Facts facts, Hash trusted_facts
+  def process_facts(request_data)
+    if request_data['facts'].nil?
+      facts = get_facts_from_pdb(request_data['certname'], request_data['environment'])
+    else
+      facts_from_request = request_data['facts']
+      facts_from_request['name'] = request_data['certname']
+      facts = Puppet::Node::Facts.from_data_hash(facts_from_request)
+    end
+
+    fact_values = if facts.nil?
+                    {}
+                  else
+                    facts.sanitize
+                    facts.to_data_hash['values']
+                  end
+
+    # Pull the trusted facts from the request, or attempt to extract them from
+    # the facts hash
+    trusted_facts = if request_data['trusted_facts'].nil?
+                      fact_values['trusted'].nil? ? {} : fact_values['trusted']
+                    else
+                      request_data['trusted_facts']['values']
+                    end
+
+    return facts, trusted_facts
+  end
+
+  def get_facts_from_pdb(nodename, environment)
+    if Puppet[:storeconfigs_backend] == :puppetdb
+      Puppet::Node::Facts.indirection.find(nodename, :environment => environment)
+    else
+      # How should this be surfaced? Seems like we could maybe do better than a 500, unless
+      # that's accurate?
+      raise(Puppet::Error, "PuppetDB not configured, please provide facts with your catalog request.")
+    end
+  end
+
+  # Initialize our server fact hash; we add these to each client, and they
+  # won't change while we're running, so it's safe to cache the values.
+  def set_server_facts
```

Comment:
Maybe put in a comment to reference where this method comes from in the [puppet compiler](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/catalog/compiler.rb#L390)?